Apps/Src/SendDisplay.c

```diff
@@ -1,30 +1,296 @@
+				.compOrCmd = (char*)compStrings[comp],
+				.attr = "en",
+				.op = "=",
+				.numArgs = 1,
+				.argTypes = {true},
+				{
+					{.num=val}
+				}
+			};
+			return Display_PutNext(toggleCmd) && Display_Refresh();
+		}
+		else{
+			Display_Cmd_t visCmd = {
+				.compOrCmd = "vis",
+				.attr = NULL,
+				.op = NULL,
+				.numArgs = 2,
+				.argTypes = {false,true},
+				{
+					{.str=(char*)compStrings[comp]},
+					{.num=val}
+				}
+			};
+			return Display_PutNext(visCmd);
+		}
+	}
+	// For components that have a non-boolean value
+	else if(comp > MOTOR){
+		Display_Cmd_t setCmd = {
+			.compOrCmd = (char*)compStrings[comp],
+			.attr = "val",
+			.op = "=",
+			.numArgs = 1,
+			.argTypes = {true},
+			{
+				{.num=val}
+			}
+		};
+		return Display_PutNext(setCmd);
+	}
+	else{
+		return ERROR;
+	}
+	return SUCCESS;
+}
+
+ErrorStatus Display_SetPage(Page_t page){
+	Display_Cmd_t pgCmd = {
+		.compOrCmd = "page",
+		.attr = NULL,
+		.op = NULL,
+		.numArgs = 1,
+		.argTypes = {true},
+		{
+			{.num=page}
+		}
+	};
+	return Display_PutNext(pgCmd);
+}
+
+/* WRAPPERS */
 
-    Display_SetMainView(); // Make sure we're in the main view first
-    Display_CruiseEnable(OFF);
-    Display_SetGear(OFF, OFF);
-    Display_SetLight(CTRL_FAULT,OFF);
+ErrorStatus Display_SetSOC(uint8_t percent){	// Integer percentage from 0-100
+	return Display_SetComponent(SOC, percent) && Display_Refresh();
+}
+
+ErrorStatus Display_SetSBPV(uint32_t mv){
+	return Display_SetComponent(SUPP_BATT, mv/100) && Display_Refresh();
+}
+
+ErrorStatus Display_SetVelocity(uint32_t mphTenths){
+	// units of .1 mph
+	return Display_SetComponent(VELOCITY, mphTenths);
+}
+
+ErrorStatus Display_SetAccel(uint8_t percent){
+	return Display_SetComponent(ACCEL_METER, percent);
+}
 
-    for(int i=0; i<6; i++)
-        Display_SetLight(i, OFF);
+ErrorStatus Display_SetArray(bool state){
+	return Display_SetComponent(ARRAY, state);
+}
+
+ErrorStatus Display_SetMotor(bool state){
+	return Display_SetComponent(MOTOR, state);
+}
+
+ErrorStatus Display_SetGear(TriState_t gear){
+	// Even though this isn't a three stage component, i'm still using the same enum because i'm lazy
+	return Display_SetComponent(GEAR, (uint32_t)gear) && Display_Refresh();
+}
+
+ErrorStatus Display_SetRegenState(TriState_t state){
+	return Display_SetComponent(REGEN_ST, (uint32_t)state) && Display_Refresh();
+}
+
+ErrorStatus Display_SetCruiseState(TriState_t state){
+	return Display_SetComponent(CRUISE_ST, (uint32_t)state) && Display_Refresh();
+}
+
+ErrorStatus Display_SetLeftBlink(bool state){
+	return Display_SetComponent(LEFT, state);
+}
+
+ErrorStatus Display_SetRightBlink(bool state){
+	return Display_SetComponent(RIGHT, state);
+}
+
+ErrorStatus Display_SetHeadlight(bool state){
+	return Display_SetComponent(HEAD, state);
+}
+
+/**
+ * @brief Loops through the display queue and sends all messages
+ */
+void Task_SendDisplay(void *p_arg) {
+    OS_ERR err;
+
+    disp_fifo_new();
```

Comment:
This should definitely be in the driver layer.

Reply:
The queue should stay in the app layer, no? I think that if the queue was in the driver layer it would basically require an application to actually work, which reduces portability. Whereas if the queue is in the app and the driver is only sending messages, then the driver can be ported to another completely separate system

Reply:
Nothing is stopping another task from calling a display function that attempts to put an item in the queue before it's initialized. Being in the driver initialization would guarantee that.

Reply:
But the queue isn't in the driver layer anymore, it's in the app

Reply:
Yes, but the functions the interface exposes call the set component function, which calls send next, which attempts to push stuff into the fifo. If the fifo isn't yet initialized (which might happen if Send Display hasn't started executing yet) then bad stuff happens.

Reply:
Right, so if the driver doesn't have the fifo, but the fifo needs to be initialized before any apps, then should a SendDisplay_Init be created and run along with driver initializations?

Reply:
If you really want to keep those calls in SendDisplay, then there should probably be a separate function to initialize the display fifo, similar to CAN_Queue_Init; then just put it in the Init task along with all of the other driver initialization calls.